File: Data Structures & Algorithms/binary-tree-preorder-traversal/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
   public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> o;
        TreeNode* curr = root;
        while (curr) {
            if (!curr->left) {
                o.push_back(curr->val);
                curr = curr->right;
            } else {
                TreeNode* pred = curr->left;
                while (pred->right != nullptr && pred->right != curr) {
                    pred = pred->right;
                }
                if(pred->right != curr){
                    o.push_back(curr->val);
                }
                if (pred->right == nullptr) {
                    pred->right = curr;
                    curr = curr->left;
                } else {
                    pred->right = nullptr;
                    curr = curr->right;
                }
            }
        }
        return o;
    }
};

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        idx_map = {val: i for i,val in enumerate(inorder)}
        self.pre_idx = 0;

        def helper(left,right):
            if left > right:
                return None
            root_val = preorder[self.pre_idx]
            self.pre_idx += 1

            mid = idx_map[root_val]
            node = TreeNode(root_val)
            node.left = helper(left,mid - 1)
            node.right = helper(mid+1,right)
            return node
        return helper(0,len(preorder) - 1)